### PR TITLE
Fix #102

### LINKED ISSUE #102: nnor-onnx: Lack of a check in MatMul optimization
https://github.com/Idein/nnoir/blob/750d27a732ccac735fdef33c34d29d7e0e6f3ee3/nnoir-onnx/nnoir_onnx/operators/mat_mul.py#L10-L22

Sometimes Matmul can be reduced to Linear, but these lines seems  lack of some necessary checks.

### PR DESCRIPTION
**Pull request by nomaddo**
This PR fix the bug of  #102.

### CODE REVIEW

File: nnoir-onnx/test/test_matmul.py [hunk 1]
```diff
@@ -37,7 +37,6 @@ def create_onnx(self) -> onnx.ModelProto:
     MatMulTester({"x": x}, outputs).run()
 
 
-@pytest.mark.xfail()
```
Comment (nomaddo) [inline nnoir-onnx/test/test_matmul.py:40]: Now, we pass the test case.

File: nnoir-onnx/nnoir_onnx/operators/mat_mul.py [hunk 2]
```diff
@@ -9,7 +18,7 @@ def __init__(self, node, *args):
 
     def to_function(self, env, constants):
         [x, W] = self.node.input
-        if W in constants:
+        if W in constants and len(constants[W].shape) == 2 and np.prod(env[x].shape[1:]) == constants[W].shape[1]:
```
Comment (notogawa) [inline nnoir-onnx/nnoir_onnx/operators/mat_mul.py:21]: MatMul should not be "chainer like" Linear.

Comment (nomaddo) [reply nnoir-onnx/nnoir_onnx/operators/mat_mul.py:21]: What do you mean?

Comment (notogawa) [reply nnoir-onnx/nnoir_onnx/operators/mat_mul.py:21]:
```
>>> x.shape
(2, 3, 4, 5)
>>> W.shape
(5, 6)
>>> np.matmul(x,W).shape
(2, 3, 4, 6)
```

Comment (nomaddo) [reply nnoir-onnx/nnoir_onnx/operators/mat_mul.py:21]: I cannot catch what you want to say.
Your example is not sufficient the condition check pointed by you.

Can you clarify what you want to say in words.

Comment (notogawa) [reply nnoir-onnx/nnoir_onnx/operators/mat_mul.py:21]: Sorry, my example is confusing.

At first, you should add and try this test case.
```python
def test_matmul_06():
    '''
    opset version >= 9
    '''

    shape = (3, 1, 1)
    w_shape = (1, 1)
    out_shape = (3, 1, 1)

    class MatMulTester(Base):
        def create_onnx(self) -> onnx.ModelProto:
            node = make_node("MatMul", inputs=["x", "W"], outputs=["y"])
            inputs = [info("x", TensorProto.FLOAT, shape)]
            outputs = [info("y", TensorProto.FLOAT, out_shape)]

            init_W = from_array(np.random.rand(*w_shape).astype(np.float32), "W")

            graph = make_graph([node], "add_graph", inputs, outputs, initializer=[init_W])
            model = make_model(graph)
            return model

    x = np.random.rand(*shape).astype(np.float32)

    outputs = ["y"]
    MatMulTester({"x": x}, outputs).run()
```

`W in constants and len(constants[W].shape) == 2 and np.prod(env[x].shape[1:]) == constants[W].shape[1]` is wrong condition. We can convert MatMul(x, W) to single Linear(x, W.T) when x.ndim == W.ndim == 2.

Comment (nomaddo) [reply nnoir-onnx/nnoir_onnx/operators/mat_mul.py:21]: Thx. I will fix my code.

File: nnoir-onnx/test/test_matmul.py [hunk 3]
```diff
@@ -167,3 +166,57 @@ def create_onnx(self) -> onnx.ModelProto:
...
+            model = make_model(graph)
+            return model
+
+    x = np.random.rand(*shape).astype(np.float32)
+
+    outputs = ["y"]
+    MatMulTester({"x": x}, outputs).run()
+
+
+def test_matmul_07():
```
Comment (notogawa) [inline nnoir-onnx/test/test_matmul.py:198]: Dup of `test_matmul_00`?

Comment (nomaddo) [reply nnoir-onnx/test/test_matmul.py:198]: That's right...

Comment (nomaddo) [reply nnoir-onnx/test/test_matmul.py]: 1b4fe3d

**Review by notogawa — APPROVED** [review]
LGTM

**Pull request merged**